Fixes: #14606 - For Cable API add PK for cable terminations to serializer

DanSheps opened the pull request:
### Fixes: #14606

* Add pk field for cable termination
* Add test

DanSheps committed:
For Cable API add PK for cable terminations to serializer

jeremystretch commented:
This one is probably a bit stickier than I realized in the original bug report.

We should really be using a nested CableTerminationSerializer here, omitting its `cable` and `cable_end` fields (as they're redundant in this context). That would address the primary issue.

However, the catch is that the generic serializer currently in use returns the fields `object_type`, `object_id` and `object`, whereas CableTerminationSerializer returns `termination_type`, `termination_id`, and `termination`. Swapping out the serializer would be a breaking change.

We _could_ address this in v4.1, however I'm hesitant to slip in a breaking change so late in the beta period. For now, we could settle with just adding the `id` field to GenericObjectSerializer, and capture this as a potential change for v4.2.

jeremystretch commented on netbox/dcim/api/serializers_/cables.py:
```diff
@@ -16,14 +16,27 @@
 )
 
 
+class GenericObjectTerminationSerializer(GenericObjectSerializer):
```
This doesn't require a new serializer: We can just add the `id` field to GenericObjectSerializer.

DanSheps replied:
Unfortunately since a_terminations/b_terminations is a property function that includes the actual termination (Interface/CircuitTermination/etc) and not the CableTermination it won't be the proper `id` (it will be the interface id for example, not the cable termination id).

This is why I opted to go this route, as as you mention, it would be a breaking change.  Perhaps we should aim for 4.1 instead and I just rebase and swap the serializer.

It may also make more sense to maintain this a_terminations/b_terminations and add a generic "terminations" which operates on `Cable.terminations` which is I think where it would be appropriate to use the CableTerminations serializer.

jeremystretch commented on netbox/dcim/tests/test_api.py:
```diff
@@ -2038,6 +2038,14 @@ def setUpTestData(cls):
             },
         ]
 
+    def test_cable_termination_pk_exist(self):
```
This test probably isn't necessary.

DanSheps committed:
Temporary work to swap in serializer

DanSheps committed:
Merge branch 'feature' into 14606-add-pk-to-cable_api_terminations

DanSheps commented:
* Rebased to feature
* Modified serializer to add `terminations` related objects

Let me know if you would rather replace `a_terminations` / `b_terminations` (will be non-trivial as they currently only return the Terminated Objects (Interface, etc)

DanSheps commented on netbox/dcim/api/serializers_/cables.py:
```diff
@@ -20,24 +20,8 @@
-        ]
-        brief_fields = ('id', 'url', 'display', 'label', 'description')
-
+class TracedCableSerializer(serializers.ModelSerializer):
```
```suggestion
class TracedCableSerializer(BaseModelSerializer):
```

DanSheps committed:
Fix merge mistake

DanSheps commented on netbox/dcim/api/serializers_/cables.py:
```diff
@@ -20,24 +20,8 @@
-
-
 class TracedCableSerializer(BaseModelSerializer):
+
```
```suggestion
```

DanSheps committed:
Update netbox/dcim/api/serializers_/cables.py

jeremystretch commented:
This is going to be more involved than I originally anticipated, because the `a_terminations` and `b_terminations` attributes of a cable instance are populated with the actual terminating objects, rather than CableTermination instances. We don't want to just tack these on in addition to the existing attributes because that would be redundant and inefficient. We'll have to explore our options, which likely include tweaking the underlying querysets for the Cable model.

jeremystretch commented:
Appreciate your work on this @DanSheps. I'm going to close it out for now; let's re-evaluate our options for a potential change in v4.2.
